### DISCUSSION TITLE
Built .apk links not working

### DISCUSSION THREAD

**Discussion by marcelo-novapaytech** [codemagic-ci-cd/codemagic-docs#2550] [Q&A] [answered]
### If you are looking for help, please confirm the following...

- [X] I confirm I have searched the [Docs](https://docs.codemagic.io), [Codemagic Sample Projects](https://github.com/codemagic-ci-cd/codemagic-sample-projects), and GitHub [Discussions](https://github.com/orgs/codemagic-ci-cd/discussions).
- [X] I confirm the software versions on my local computer match the ones in CI environment (e.g. Xcode version, Gradle version). See [here](https://docs.codemagic.io/troubleshooting/common-android-issues/#builds-work-locally-but-fail-on-codemagic)
- [X] I confirm I am able to do a fresh clone of the repository and build the project on my local computer.

### Which mobile framework are you using?

Flutter (Dart)

### Steps to reproduce

My builds posts .apk links to Slack.

### Expected results

It was working fine until yesterday.

### Actual results

But today's build gives us an error message when clicked:

```
{"error":"FORBIDDEN","message":"The server could not verify that you are authorized to access the URL requested. You either supplied the wrong credentials (e.g. a bad password), or your browser doesn't understand how to supply the credentials required."}
```

It works only if I download this while logged into my CodeMagic account. 

Is this a new configuration, or a bug?

### Build id (optional)

664e405803e36d6c4996d8c1

**Comment by dtrdic** [comment, marked as answer]
Hello @marcelo-novapaytech the error you're encountering is expected behavior. The built .apk links provided by Codemagic are restricted to team members and require authentication. This has always been the case to ensure the security of your build artifacts.

To share the builds outside your team, you need to convert the link to a public download URL. Here's a basic outline of how you can achieve that
```
  publishing:
    scripts:
      - name: Publish to Slack
        script: |
          # Fetch the artifact URL for the .apk file
          ARTIFACT_URL=$(echo $CM_ARTIFACT_LINKS | jq -r '.[] | select(.name | endswith(".apk")) | .url')

          # Set the expiration timestamp for the public URL (3 days from now)
          expiration_timestamp=$(date -v+3d +%s)

          # Create a public download URL for the artifact
          response=$(curl -s -H "Content-Type: application/json" \
                          -H "x-auth-token: $CODEMAGIC_API_TOKEN" \
                          -d "{\"expiresAt\": $expiration_timestamp}" \
                          -X POST $ARTIFACT_URL/public-url)
          
          public_url=$(echo $response | jq -r '.url')

          # Post the public URL to Slack
          curl -s -X POST $SLACK_WEBHOOK_URL \
              -H 'Content-type: application/json' \
              --data-raw '{
                  "channel": "#testprivatechannel",
                  "attachments": [
                      {
                          "blocks": [
                              {
                                  "type": "section",
                                  "text": {
                                      "type": "mrkdwn",
                                      "text": "Build with build id <https://codemagic.io/app/'"$CM_PROJECT_ID"'/build/'"$CM_BUILD_ID"'>"
                                  }
                              },
                              {
                                  "type": "divider"
                              },
                              {
                                  "type": "section",
                                  "text": {
                                      "type": "mrkdwn",
                                      "text": "*Android Artifact Link:* <'"$public_url"'|Download>"
                                  }
                              }
                          ]
                      }
                  ]
              }'
```
Check this links for reference
https://docs.codemagic.io/rest-api/artifacts/
https://docs.codemagic.io/knowledge-others/slack-api-integration/
